### ISSUE TITLE
Randomized textures (sprite strip)

### ISSUE THREAD

**Issue by KenneyNL** [minetest/minetest#2746]
In a same way that nodes are currently animated (sprite strip) it could also function to give nodes a random appearance. Upon placing the node, a random frame of the stripe strip will be selected. The random seed could be done using a noise.

This would eliminate the duplication of nodes to achieve random node textures.

**Comment by celeron55** [comment]
(Review for bounty: OK)

**Comment by C1ffisme** [comment]
I made an issue like this at some point, I guess it got lost.

**Comment by TeTpaAka** [comment]
#2433 is related. I think this was the one @C1ffisme meant.

**Comment by KenneyNL** [comment]
#2433 would still clutter the texture folder though with multiple alternatives of the same texture, I think having them all in 1 texture file is a bit more handy.

**Comment by C1ffisme** [comment]
I dunno, it may help some people with texture packs, depending on what they are trying to do.

We could add both as an alternative for people.

**Comment by ExcaliburZero** [comment]
This feature sounds like it could be useful.

I agree with KenneyNL that it would probably be better to use sprite strips rather than individual files, as it would be more like the current system for animated nodes.

**Comment by Wuzzy2** [comment]
Some questions arise:

How and when should the sprite be selected?
And how and where will the selected sprite be stored?

Should the sprite be allowed to change on a game restart?
Should it be equally for all clients or does every client determine the sprite by itself?

**Comment by KenneyNL** [comment]
Some things I can answer from a game designer perspective;

1. Upon placing/constructing the node.
2. Using some sort of noise algorithm? No idea on that.
3. Nope, random textures should stay the same.
4. It should be equal for everyone.

**Comment by 4aiman** [comment]
2. Wasn't the increase of additional info to store amongst the reasons to decline coloured light sources? (I'm only asking)

**Comment by PilzAdam** [comment]
@4aiman there is no need to store this extra information; it can be pseudo-randomly generated each time the mapblock mesh is updated.

**Comment by C1ffisme** [comment]
@PilzAdam Now there's a good idea, give the man a medal!

Okay, so it's kindof hard to do that over the internet.

In fact, we could just use the seed provided by the world already, seeing as it's used for the rest of mapgen. That might work.

**Comment by 4aiman** [comment]
@PilzAdam so the textures will "dance" on every update?
I thought the original idea was to preserve the selected texture over time.

**Comment by cheapie** [comment]
@4aiman I think the idea was to have the seed be based off of the mapblock position, so that the "random" numbers will be the same every time for any given mapblock.

**Comment by ExcaliburZero** [comment]
I think as C1ffisme and cheapie mentioned, if the random texture would be based on either just the node position or both the node position and the map seed (for a little bit more randomness), then the texture chosen should appear to be "random" while still remaining consistent.

**Comment by 4aiman** [comment]
Oh... now I get it :)
So, I won't be able to really choose the texture (i.e. every time I place a node in a certain position it will use the same texture)?
I'd rather want to be able to control the pattern by choosing a texture somehow.

I acknowledge that the initial problem raised in current issue will be solved and fully covered by PilzAdam's solution, but still.. 

Also, there's a separate issue about the connected textures. 
It would be nice to integrate those into this solution and create a way to select the chosen texture (i.e. override the pseudo-randomly chosen one).

How hard would it be to add a way to use  metadata to store the number of the texture?
A certain pre-defined field name  can be used.
Will that kill performance?

**Comment by ExcaliburZero** [comment]
@4aiman I agree that it would probably be good to have a way to set the texture of the node, as that would open up a lot of other possibilities such as connected nodes, nodes that change their appearance (like furnaces, battery boxes), etc. In the case of a random texture a random number could just be selected on create to determine the random texture.

However, I'm not sure how well that would end up working in terms of storing the metadata.

**Comment by 4aiman** [comment]
@ExcaliburZero I'm not insisting on anything.
I'm just trying to find a way to suppress the growth of the size of a mapblock. :)
All because initial coloured light was declined due to doubling (or so I was told) the size of a mapblock.

**Comment by sofar** [comment]
This can be implemented using a simple random number generator that is seeded for each surface with e.g. (x + z <<8 + y <<16 + face). For an example how to make randomness consistent across mesh/map updates look at my random_xz patch or the "plant fun" patch which has this element in it. (PR 3923)

**Comment by Ryan-Nolan** [comment]
Would storing a single number make such a big diffrence to map blcok size? i have very little understanding of that at the moment. but it does seem like an extremly small amount of data..

this feature would also mean less nodes being created due to the many nodes that create 2 of each for purposes of having diffrent looks etc.. 

Grass textured nodes would especially benefit from this, being able to loose that repetative look. thats almost imposible to get rid off without leaving it looking plain.

If it really bugs people, maybe providing 2 versions of this would be better:
1) that does the random picking of a sprite without actualy storing it.(in other words it would be diffrent everytime the node is loaded.)
2) non-random that defaults to the first sprite but is changeable in the node defintion same way in the same way we fould change infotext.

so in essence you would use method 1 for things like grass and other large surfaces and you would use method 2 for things like furnaces..

getting the best of both worlds. ?

**Comment by sofar** [comment]
Having a method like ```random_xz``` had would require no map storage.

**Comment by Ryan-Nolan** [comment]
@sofar thats a good point. but where would it store it. In a file of its own?
And im presuming if it doesnt store it at all.. that would mean its no longer random just diffrent for each pos?

Either way it sounds like a good solution. Im just curios :)

**Comment by sofar** [comment]
it can't be truly random, as the mesh for each block gets regenerated regularly and you don't want the texture of your grass to change just because someone placed a dirt block somewhere.

A fixed seed works well. This is all client-side as well.

**Comment by rubenwardy** [comment]
Can't you just use a RNG with a seed of (x*100+y*10+z+mapgenseed) % m

**Comment by sofar** [comment]
@rubenwardy that's what that is.

**Comment by paramat** [comment]
:-1: 
Clever idea but i feel it is over-complex to move towards needing multiple textures per node face. I know it would be 'optional' but i don't think over-complex non-essential ideas should be encouraged or accepted in Minetest even as options. With our dev shortage we need to keep things simple and resist overly fancy stuff. Our worlds look fine with repeating nodes. For those used to this feature in Minecraft, too bad, don't let that make you see Minetest as wprse.

**Comment by C1ffisme** [comment]
@paramat I'm not sure you see the real advantage of this. This would mean you could have randomized bookshelves, or other decor, without needing to register several nodes.

I think that Minetest, the engine, should support people who want features like this.

**Comment by sofar** [comment]
I wouldn't do a sprite strip. We already have this exact mechanism for sounds: e.g. if you request "sound_place" you can get any of "sound_place.ogg" "sound_place.1.ogg" "sound_place.2.ogg" "sound_place.3.ogg". This sounds much more generic and could be used for any texture, including normal maps and overlays.

**Comment by cheapie** [comment]
How about a method to override textures through node metadata? That would
come in handy for my rgblightstone mod that currently registers 4096 nodes.
I'd be able to replace that with just one, and also add more colors at the
same time.

**Comment by C1ffisme** [comment]
@cheapie That would be nice. But while were at it, we should really make all parts of a node be override-able by metadata.

**Comment by sofar** [comment]
@cheapie that's nice but it would mean that mods have to touch *every* node and attach metadata, which is expensive and adds storage space.

I see that as a solution for another problem, and so it shouldn't be used to solve a different problem.

**Comment by cheapie** [comment]
Nodes would probably still have textures in the definition, and the texture
override metadata would only have to be present if the texture is, well,
overridden.

**Comment by 4aiman** [comment]
@cheapie
Despite I was  the one who wanted to override textures using metadata, I think that's not such a great idea as of now.

Getting metadata before drawing a node would mean reading data from a server.
I don't know whether locally saved map or a server one will be accessed, though.

**Comment by paramat** [comment]
> I'm not sure you see the real advantage of this. This would mean you could have randomized bookshelves, or other decor, without needing to register several nodes.

Yeah i do see use for the idea, but it can be done in mods with extra nodes, basic Minetest just doesn't need this enough to justify making critical code more complex. Randomisation is a more specialist thing so a mod thing, it should not burden the basic game.

**Comment by C1ffisme** [comment]
@paramat 
> but it can be done in mods with extra nodes

> C1ff said:
 **without needing to register several nodes.**

Adding four extra nodes that do the same thing, have the same hardness, etc, but are different only in texture just means that four more nodes are being loaded at loading time. 

> forgetting the _core priciples_ of Minetest: simple and lightweight.

We have core principles? Because last I checked, we can't even sort out [our development direction](https://github.com/minetest/minetest/issues/3476).

(Yes, I see that we added a `CONTRIBUTING.md`. That has to do with code styling, not development goals.)

I'm pretty sure that you are the only one who has said anything about Minetest being 'Simple and Lightweight'.

**Comment by rubenwardy** [comment]
How many mapgens do we have again?

**Comment by sofar** [comment]
> Yeah i do see use for the idea, but it can be done in mods with extra nodes, basic Minetest just doesn't need this enough to justify making critical code more complex. Randomisation is a more specialist thing so a mod thing, it should not burden the basic game.

It can be done client-side, without any server peformance impact, and almost no client side rendering impact.

Why do we want people to make hundreds of new nodetypes again?

**Comment by Zeno-** [comment]
Personally I think this is a great idea

**Comment by Ryan-Nolan** [comment]
I agree, @paramat  I love the work you have done. the mapgen V7   is awsome..... but based on your anology rivers and mountains would be classed as too complex... so they shouldnt be threre?

This has nothing todo with wether minecraft uses it or not, this is based on what minetest devs/users want and overall it looks like it is wanted my the majority of those that have commented both in this issue and  others like it.

This idea will improve both the overall visual looks of the game and the quality of mods that use multiple nodes to create the same effect.

**Comment by hunter688** [comment]
Screenshot from Eloraam's Minecraft clone:

https://twitter.com/therealeloraam

![tiling](https://cloud.githubusercontent.com/assets/1410134/15650449/eef51fa2-2678-11e6-9593-33a79e1a80e9.png)

will be awesome in the Minetest

**Comment by yyt16384** [comment]
It should probably be done in a way that allows texture packs to supply more or less textures.

**Comment by paramat** [comment]
I don't thnk it improves appearence by much, if at all, the screenshots above show a badly designed texture in the first example so those are misleading.
Minetest worlds look fine as they are with repeating textures, i obsessively watch Minecraft videos and have never thought their use of randomised textures improves the appearence. After all a voxel world is all about identical cubes.
I still don't think it's worth the extra complexity in meshgen, and moves towards an expectation of multiple textures per node makes modding and texturing more difficult.

**Comment by Fixer-007** [comment]
I guess with a nice balanced texture you don't need to randomise anything. On those screenshots above we have both bad texture and North Korea like design :)

**Comment by 0-afflatus** [comment]
I think it's a great idea, but it does depend on the implementation being simple and general enough. I could live without it just as easily.

**Comment by paramat** [comment]
Duplicate of #2433

**Comment by paramat** [comment]
Re-opening as this has more discussion than #2433 and discusses that method too.
Closing #2433 as the duplicate even though that came first.

**Comment by tobyplowy** [comment]
I think this should get added it just looks really good especially with higher resolution texture packs (it can also really look good on Lower res texture packs like pixel perfection a 16x16 pack) just look at this picture http://u.cubeupload.com/XSSheep/20160625005154.png most notable is the cobblestone on the houses and the path blocks with the little bits of grass on it.

**Comment by tobyplowy** [comment]
@paramat i think this should get added it just looks really good especially with higher resolution texture packs (it can also really look good on Lower res texture packs like pixel perfection a 16x16 pack) just look at this picture http://u.cubeupload.com/XSSheep/20160625005154.png most notable is the cobblestone on the houses and the path blocks with the little bits of grass on it.

**Comment by paramat** [comment]
Consider sofar's idea:
> I wouldn't do a sprite strip. We already have this exact mechanism for sounds: e.g. if you request "sound_place" you can get any of "sound_place.ogg" "sound_place.1.ogg" "sound_place.2.ogg" "sound_place.3.ogg". This sounds much more generic and could be used for any texture, including normal maps and overlays.

**Comment by numberZero** [comment]
So, we have several options:
1. `param2`-based texture choosing. Fully controllable, suitable for various usage patterns. Requires `param2` to be set in the map (thus increasing compressed size), so not very good for large solid chunks.
2. Random texture choosing. Suitable for large areas to reduce visual repetitiveness.
3. Large textures (#6105). Designed for large areas for the same reason.

I’d prefer (1) and (3) (for different use cases), but have nothing against (2) as yet another option. We have random Y-offset for plantlike, after all.

**Comment by Ryan-Nolan** [comment]
Whilst we are on this subject... is there a reason why we use different texture files for each face instead of 1 texture that wraps the node? i would think the latter would be more efficient

**Comment by tobyplowy** [comment]
@Ryan-Nolan I think that's not a very good idea 
It's not More efficient for artists (maybe for programmers it is) 
And it would also make minetest game bigger (Filesize)

Warning I'm just stating my opinion I'm not negative :)

**Comment by NathanSalapat** [comment]
@Ryan-Nolan If we used one file with each face a texture would have to be six times larger if every face was the same, as it would have to include the same texture six times. Now it's possible that somehow it could be coded to read differently depending on the texture size, but then if you wanted to have two faces different you'd still have to use a texture with all six faces on it. Like @tobyplowy said not very efficient for artists and it would lead to bloat.

**Comment by rubenwardy** [comment]
It also wouldn't help with performance at all. The Engine doesn't render all sides of a cube. A related optimisation which would be worth doing is texture atlasing - this existed in previous persions (until 0.4.2 or so) but was buggy due to a bad implementation. I believe this could have a good performance improvement by not switching texture context when rendering.

**Comment by tacotexmex** [comment]
sofar is on point with [this comment](#issuecomment-215806848): We already have it for sounds, why not textures too? There's plenty of texture packs with multiple alternates that can support it immediately.

Currently I have to register 5-10 node definitions for each unique node in order to support it. And that's only when placing the node (on_place), not in mapgen at all.

A [quite horrible example](#issuecomment-222490322) shouldn't have to botch the whole issue.

**Comment by rubenwardy** [comment]
Buildaworld have a fork of irrlicht with some more graphics features, may be worth looking at that

**Comment by numberZero** [comment]
@rubenwardy Their fork looks great! It seems to have everything we need!

Of course it needs some adoption, e.g. they dropped Win32 support recently, we can’t do that. But it looks very-very promising)

**Comment by tacotexmex** [comment]
Since I disliked the previous example I felt obliged to come up with a better one. Here's Pixel Perfection with 3–10 alternating textures per node. My favorite is the stonebrick wall.

![screenshot_20180414_132339](https://user-images.githubusercontent.com/25328789/38770115-f67929e8-400d-11e8-8f9a-c708f02032e4.png)
![screenshot_20180414_132342](https://user-images.githubusercontent.com/25328789/38770116-f6988716-400d-11e8-975b-c9c26e579da0.png)

**Comment by 4w** [comment]
I implemented that in Lua without any need for texture or engine changes. It works, but it’s hacky, though :smile:

![dirt](https://user-images.githubusercontent.com/30605057/38771072-b0f97f12-401c-11e8-81c4-c7780a5b7cf7.gif)
![grass_sand](https://user-images.githubusercontent.com/30605057/38771073-b10f90b8-401c-11e8-8db1-0f4924f73953.gif)
![sand_only](https://user-images.githubusercontent.com/30605057/38771074-b125bf64-401c-11e8-9f9a-57b065258473.gif)

**Comment by tacotexmex** [comment]
That solution is fine if you just want random rotation of nodes, but it doesn’t work reliably for alternating textures (no other interacting mod will recognize `default:stone_2` etc).

**Comment by 4w** [comment]
> That solution is fine if you just want random rotation of nodes,

The solution is horribly hacky and after releasing it a better solution without flooding the world with metadata came in my mind :smile:

> but it doesn’t work reliably for alternating textures […]

Yes, rotating nodeboxes would break how they are intended to look, but that wasn’t my goal. If ever properly implemented this implementation should handle it correctly and not simply override/force anything.

With my hacky code it’s possible to define the facedirs for the nodes so all facedirs where the node looks bad in can be omitted while defining the allowed rotation facedirs. I do this for [grass](https://github.com/4w/redef/blob/5326705f1119f6228a364f34e40e18b2dc9497ee/redefinitions/random_nodes.lua#L10-L11) for example.

**Comment by tacotexmex** [comment]
So who's disapproving of sofar's idea of implementation? No-one? Ok, let's move forward.

### Multiple alternate textures per node, randomized each placement

Now, even though #2433 is closed in favor of this long-winded and sprawly discussion, it quite sums it up. There would be no new Lua API addition because just like how sound alternates work, the engine would at each rendering of a node look for textures called
`default_stone.2.png`, `default_stone.3.png`, `default_stone.4.png` and so on.

**Comment by numberZero** [comment]
How should it interoperate with texture combining?

**Comment by tacotexmex** [comment]
Not sure 🤔

**Comment by Ryan-Nolan** [comment]
@numberZero I think initially it should probably ignore any texture combining arguments to keep it simple. At the moment i cant imagine an example where this would be useful with texture combining. imo

**Comment by tacotexmex** [comment]
The one I think of is alternates of the dirt texture because that texture is usually overlayed with a grass texture for sides.

**Comment by Ryan-Nolan** [comment]
@tacotexmex That's a good point, i didn't think of that

**Comment by Froggo8311** [comment]
> Some things I can answer from a game designer perspective;
> 
>     1. Upon placing/constructing the node.
> 
>     2. Using some sort of noise algorithm? No idea on that.
> 
>     3. Nope, random textures should stay the same.
> 
>     4. It should be equal for everyone.

In regard to numbers 3 & 4... Maybe allow the client to randomize it? If there are a lot of randomized nodes in the world (say, implementing a randomized grass texture), it could increase the size of data storage & transfer. I could be wrong, but I can't think of a use case where randomized textures being consistent between players & clients would be useful.

**Comment by Desour** [comment]
(For choosing a random texture one can make a hash of the node pos, tile index and map seed. It won't change after digging and placing the node again, but that's fine imo.)

-----------------------

Related:
Instead of choosing a random texture from a set, one can use stochastic texture sampling.
See also: https://github.com/minetest/minetest/issues/12746#issuecomment-1262492774
Unity3D uses it for example: https://github.com/UnityLabs/procedural-stochastic-texturing#read-more

**Comment by HybridDog** [comment]
Here are some screenshots from a modified stochastic texture sampling demo with Minetest textures.

grass, snow, stone and coal block
![2022-11-19-141000_1920x1054_scrot](https://user-images.githubusercontent.com/3192173/202852800-1cc6de65-0f5f-41ea-aed3-4932267a2d22.png)
![2022-11-19-141307_652x524_scrot](https://user-images.githubusercontent.com/3192173/202852804-b8744312-3865-47c5-a0dc-f3f820f9161b.png)
![2022-11-19-141437_1920x1054_scrot](https://user-images.githubusercontent.com/3192173/202852805-9b43a319-8555-4fe8-bbf7-da67ef97b980.png)
![2022-11-19-141719_1920x1054_scrot](https://user-images.githubusercontent.com/3192173/202852807-dc920655-5426-4523-89f7-d048d5dc6011.png)

Limitation: It works badly with textures which have connected structures, e.g. bricks and cobblestone:
![2022-11-19-141825_564x376_scrot](https://user-images.githubusercontent.com/3192173/202852808-25caa5c3-8e60-4f55-9cb1-1d1fbb3d39b2.png)
![2022-11-19-141140_849x898_scrot](https://user-images.githubusercontent.com/3192173/202852803-aef34bef-d209-4cfa-9158-e6af3e253f94.png)

**Comment by HybridDog** [comment]
> Stochastic texture sampling appears to be much more complex, I do not know if it is really worth this for low resolution textures as used here

In my opinion, it gives good results with many 16x16 textures and its complexity is low enough that it can be implemented with SVG filters (https://codeberg.org/HybridDog/sts-svg). However, integrating it into Luanti is complex because, for example, Luanti does not use a linear algebra library with which the principal component analysis can be implemented with a small amount of code and the parameters for the inverse histogram transformation somehow need to be added to the data structures which contain the main histogram-transformed texture.
Here's a forum topic about my implementation attempt in Luanti: https://forum.luanti.org/viewtopic.php?t=30600

> An alternative (or addition!) would be to use a random texture offset. Instead of a 16x16 texture, we would make a 32x32 texture and render a 16x16 chunk at a random xy offset from 0 to 15 (possibly also a random rotation). But rotation may cause issue when the texture is supposed to convey lighting, and random offsets still will not look well with stones. I think the "random from fixed list of textures" is a decent compromise.

I think using a random xy offset from 0 to 15 wouldn't work because an offset texture does not tile with an offset texture of a neighbouring node.
Nonetheless, I like the idea of adding all texture alternatives into a single image file. For example, four 16x16 texture alternatives can be put in a 16x64 image file. Luanti already uses this for [animated textures](https://github.com/luanti-org/luanti/blob/78293404c74e6c920f530046bc05dddd0a616959/doc/lua_api.md#tile-animation-definition) (e.g. the water in minetest_game). While nodes with animated textures all show the same so-called frame and the frame switches over time, with random texture selection, the nodes would show a randomly selected frame and it doesn't switch over time.

> in the pull request the variant has to be chosen via param2

For randomised textures, I prefer pseudo-random texture selection since using param2 has some disadvantages:
If param2 is chosen randomly, the size of the map database may increase significantly since random numbers cannot be compressed losslessly well. Furthermore, param2 can no longer be used for something else, so stairs and slabs orientation, hardware node colouring and other features which use param2 to store information cannot be used together with the texture selection.
